shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java

```diff
@@ -898,6 +906,27 @@ void onUserLeaveHint() {
     }
   }
 
+  /**
+   * Invoke this from {@code Activity#onWindowFocusChanged()}.
+   *
+   * <p>A {@code Fragment} host must have its containing {@code Activity} forward this call so that
+   * the {@code Fragment} can then invoke this method.
+   */
+  void onWindowFocusChanged(boolean hasFocus) {
+    ensureAlive();
+    if (host.shouldDispatchAppLifecycleState() && flutterEngine != null) {
+      // This may notify in a redundant fashion with onResume and onPause. The
+      // lifecycle channel will deduplicate, but since the Android documentation
+      // says that we shouldn't rely on any particular order between lifecycle
+      // events and onWindowFocusChanged, we notify in both cases.
+      if (hasFocus) {
+        flutterEngine.getLifecycleChannel().appIsResumed();
+      } else {
+        flutterEngine.getLifecycleChannel().appIsInactive();
```

Comment:
Is there any chance that this method could get called while the activity is stopped (after `onStop()`)? If so, this would break the `paused` lifecycle state. Even more so than the comment above about checking if the activity is Android-"resumed", I think it's probably necessary to check here that the activity is at least in the Android-"started" state.

Reply:
I can't say for sure, but it seems unlikely given Android's state machine. Any app that is stopped wouldn't be receiving any focus changes (or much of anything else event related, I would think), and it should go through "onPause" first, which would seem to me to be where Android should tell any app that it has lost focus. So there might be redundant "inactive" state notifications, but we wouldn't get an `onWindowFocusChanged(true)` until right before or right after `onResume` was called again (order depending on how Android sends its events), and at that time it would be warranted.

However, as you can probably tell, I'm not positive of any of that. These sorts of questions are why I need help here from the Android experts.

Reply:
The documentation I read said you shouldn't rely on the ordering of `onWindowFocusChanged()` and other lifecycle methods, which I assume also includes `onStop()`. This makes me suspect it could be possible for the methods to be called in quick succession in order of `onPause()`-`onStop()`-`onWindowFocusChanged(false)`.

Reply:
Hmm.  I read that too, which is why I was doing deduplication instead of trying to rely on the last reported state to decide. Sigh.  OK, if it can be that far out of sync, then I guess I have no choice but to keep track of both the Android state and the focus state locally, and send the appropriate state when they meet the right criteria, and just hope that they don't get out of sync with the state that the framework thinks it is in. That should only be possible when other code besides the `LifecycleChannel` sends messages to the `flutter/lifecycle` channel, so hopefully it's pretty rare.

I added a state table to the PR description to help clarify what states would be in force when.